network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -4,86 +4,117 @@ package validator
+	ErrUnknownMessageType = errors.New("validation failed: failed to get message auth config")
+	ErrIdentityUnverified = errors.New("validation failed: could not verify identity of sender")
 )
 
-// AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
-// is part of the authorized roles list for the channel being communicated on. A node is considered
-// to be authorized to send a message if the following are true.
-// 1. The node is staked.
-// 2. The node is not ejected.
-// 3. The message type is a known message type (can be decoded with network codec).
-// 4. The message is authorized to be sent on channel.
-// 4. The sender role is authorized to send message channel.
-// 5. The sender role is authorized to participate on channel.
-func AuthorizedSenderValidator(log zerolog.Logger, channel network.Channel, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
+type validateFunc func(ctx context.Context, from peer.ID, msg interface{}) (string, error)
+
+// AuthorizedSenderValidator returns a MessageValidator that will check if the sender of a message is authorized to send the message.
+// The MessageValidator returned will use the getIdentity to get the flow identity for the sender, asserting that the sender is a staked node and not ejected. Otherwise, the message is rejected.
+// The message is also authorized by checking that the sender is allowed to send the message on the channel.
+// If validation fails the message is rejected, and if the validation error is an expected error, slashing data is also collected.
+// Authorization config is defined in message.MsgAuthConfig
+func AuthorizedSenderValidator(log zerolog.Logger, channel channels.Channel, getIdentity func(peer.ID) (*flow.Identity, bool)) validateFunc {
 	log = log.With().
 		Str("component", "authorized_sender_validator").
 		Str("network_channel", channel.String()).
 		Logger()
 
-	return func(ctx context.Context, from peer.ID, msg interface{}) pubsub.ValidationResult {
+	slashingViolationsConsumer := slashing.NewSlashingViolationsConsumer(log)
+
+	return func(ctx context.Context, from peer.ID, msg interface{}) (string, error) {
 		identity, ok := getIdentity(from)
 		if !ok {
-			log.Warn().Str("peer_id", from.String()).Msg("could not verify identity of sender")
-			return pubsub.ValidationReject
+			log.Error().Err(ErrIdentityUnverified).Str("peer_id", from.String()).Msg("rejecting message")
+			return "", ErrUnauthorizedSender
 		}
 
-		// redundant check if the node is ejected so that we can fail fast before decoding
-		if identity.Ejected {
-			log.Warn().
-				Err(fmt.Errorf("sender %s is an ejected node", identity.NodeID)).
+		msgType, err := isAuthorizedSender(identity, channel, msg)
+		switch {
+		case err == nil:
+			return msgType, nil
+		case errors.Is(err, ErrUnauthorizedSender):
+			slashingViolationsConsumer.OnUnAuthorizedSenderError(identity, from.String(), msgType, err)
```

Comment:
we should include the channel in the error logs. maybe it could go in the error returned by `IsAuthorizedSender`?

Reply:
The network channel is added to the logger before the slashing violations consumer is created [here](https://github.com/onflow/flow-go/blob/khalil/1704-masquerade-slashing-hooks/network/validator/pubsub/authorized_sender_validator.go#L37)